### PR TITLE
Fix grpc connections close

### PR DESCRIPTION
**Pull request by tomekl007**
**What this PR does**:

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [ ] Changes manually tested
- [ ] Automated Tests added/updated
- [ ] Documentation added/updated

### CODE REVIEW

File: grpc/src/main/java/io/grpc/netty/shaded/io/grpc/netty/CustomChannelFactory.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  public static EventLoopGroup boss() {
+    SharedResourcePool<EventLoopGroup> group =
+        SharedResourcePool.forResource(Utils.DEFAULT_BOSS_EVENT_LOOP_GROUP);
+    return group.getObject();
+  }
+
+  @Override
+  public ServerChannel newChannel() {
+    ServerChannel serverChannel = channelFactory.newChannel();
+    channels.add(serverChannel);
```
Comment (ivansenic) [inline grpc/src/main/java/io/grpc/netty/shaded/io/grpc/netty/CustomChannelFactory.java:49]: `channels` is non-thread safe `ArrayList`, can newChannel be called concurrently? if so please adapt to the synchronized list data type..

File: grpc/src/main/java/io/grpc/netty/shaded/io/grpc/netty/CustomChannelFactory.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+ * uses it for session creation. Once it's notified via the {@code closeFilter()} method, it closes
+ * all channels matching the predicate.
+ */
+public class CustomChannelFactory implements ChannelFactory<ServerChannel> {
+  private static final Logger logger = LoggerFactory.getLogger(CustomChannelFactory.class);
+  private final ChannelFactory<? extends ServerChannel> channelFactory;
+  private final List<Channel> channels = new ArrayList<>();
+
+  public CustomChannelFactory() {
+    this.channelFactory = Utils.DEFAULT_SERVER_CHANNEL_FACTORY;
```
Comment (ivansenic) [inline grpc/src/main/java/io/grpc/netty/shaded/io/grpc/netty/CustomChannelFactory.java:31]: pass to the `CustomChannelFactory(ChannelFactory<? extends ServerChannel> channelFactory)`, this is the only way to have this unit tested and pass mocked `ChannelFactory`

File: grpc/src/main/java/io/grpc/netty/shaded/io/grpc/netty/CustomChannelFactory.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    return serverChannel;
+  }
+
+  public void closeFilter(Predicate<Map<String, String>> headerFilter) {
+    ChannelGroup allChannels = new DefaultChannelGroup(GlobalEventExecutor.INSTANCE);
+    channels.stream()
+        .filter(
+            channel -> {
+              ProxyInfo proxyInfo = channel.attr(ProxyInfo.attributeKey).get();
+              logger.info("proxyInfo:" + proxyInfo);
```
Comment (ivansenic) [inline grpc/src/main/java/io/grpc/netty/shaded/io/grpc/netty/CustomChannelFactory.java:59]: `info` level? should be debug or? same below?

File: grpc/src/main/java/io/grpc/netty/shaded/io/grpc/netty/CustomChannelFactory.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+   * close the underlying connection gracefully.
+   */
+  private void writeToGroup(ChannelGroup allChannels) {
+    logger.info("Closing channels, number of channels: {} ", allChannels.size());
+    GracefulServerCloseCommand streamClosed = new GracefulServerCloseCommand("Stream closed");
+    try {
+      allChannels.writeAndFlush(streamClosed).get();
+      allChannels.close().get();
+    } catch (InterruptedException | ExecutionException e) {
+      throw new RuntimeException("problem when write GracefulServerCloseCommand", e);
```
Comment (ivansenic) [inline grpc/src/main/java/io/grpc/netty/shaded/io/grpc/netty/CustomChannelFactory.java:80]: do we actually wanna throw this back?

File: grpc/src/main/java/io/grpc/netty/shaded/io/grpc/netty/ProxyInfo.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+public class ProxyInfo {
+  public static final AttributeKey<ProxyInfo> attributeKey = AttributeKey.valueOf("PROXY");
+
+  public final InetSocketAddress destinationAddress;
+  public final InetSocketAddress sourceAddress;
+  private final Map<String, String> headers;
+
+  public ProxyInfo(InetSocketAddress destinationAddress, InetSocketAddress sourceAddress) {
+    this.destinationAddress = destinationAddress;
+    this.sourceAddress = sourceAddress;
```
Comment (ivansenic) [inline grpc/src/main/java/io/grpc/netty/shaded/io/grpc/netty/ProxyInfo.java:20]: we keep this only for purposes of `toString()`?

File: grpc/src/main/java/io/stargate/grpc/impl/EventNotifier.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+  @Override
+  public void onClose(Predicate<Map<String, String>> headerFilter) {
+    if (headerFilter != null) {
+      customChannelFactory.closeFilter(headerFilter);
+    }
+  }
+
+  @Override
+  public void onCreateKeyspace(String keyspace, Predicate<Map<String, String>> headerFilter) {}
```
Comment (ivansenic) [inline grpc/src/main/java/io/stargate/grpc/impl/EventNotifier.java:26]: would be nice to have `EventListenerWithChannelFilterAdapter` and then you don't have to implement all methods from the `EventListenerWithChannelFilter`?

**Review by ivansenic — COMMENTED** [review]
Uff, I just realized this is a draft.. Anyway would be a shame that comments are flushed, so few suggestions..

**Comment by mpenick** [comment]
This was an attempt to provide a way to close long-lived gRPC/HTTP/2 connections when a tenant is removed from a CNDB coordinator. It was unsuccessful so far.

**Pull request closed without merging**